plugins/inputs/xtremio/xtremio.go

```diff
@@ -0,0 +1,401 @@
+		"ssds_read_bandwidth":   data.Content.ReadBandwidth,
+		"ssds_num_bad_sectors":  data.Content.NumBadSectors,
+	}
+
+	acc.AddFields("xio", fields, tags)
+}
+
+func (xio *XtremIO) gatherVolumes(acc telegraf.Accumulator, url string, wg *sync.WaitGroup) {
+	defer wg.Done()
+	resp, err := xio.call(url)
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	data := Volumes{}
+	err = json.Unmarshal([]byte(resp), &data)
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	tags := map[string]string{
+		"guid":     data.Content.GUID,
+		"sys_name": data.Content.SysName,
+		"name":     data.Content.Name,
+	}
+	fields := map[string]interface{}{
+		"volumes_read_iops":            data.Content.ReadIops,
+		"volumes_write_iops":           data.Content.WriteIops,
+		"volumes_read_latency":         data.Content.ReadLatency,
+		"volumes_write_latency":        data.Content.WriteLatency,
+		"volumes_data_reduction_ratio": data.Content.DataReductionRatio,
+		"volumes_provisioned_space":    data.Content.ProvisionedSpace,
+		"volumes_used_space":           data.Content.UsedSpace,
+	}
+
+	acc.AddFields("xio", fields, tags)
+}
+
+func (xio *XtremIO) gatherXMS(acc telegraf.Accumulator, url string, wg *sync.WaitGroup) {
+	defer wg.Done()
+	resp, err := xio.call(url)
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	data := XMS{}
+	err = json.Unmarshal([]byte(resp), &data)
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	tags := map[string]string{
+		"guid":    data.Content.GUID,
+		"name":    data.Content.Name,
+		"version": data.Content.Version,
+		"xms_ip":  data.Content.IP,
+	}
+	fields := map[string]interface{}{
+		"xms_write_iops":               data.Content.WriteIops,
+		"xms_read_iops":                data.Content.ReadIops,
+		"xms_overall_efficiency_ratio": data.Content.EfficiencyRatio,
+		"xms_ssd_space_in_use":         data.Content.SpaceUsed,
+		"xms_ram_in_use":               data.Content.RAMUsage,
+		"xms_ram_total":                data.Content.RAMTotal,
+		"xms_cpu_usage_total":          data.Content.CPUUsage,
+		"xms_write_latency":            data.Content.WriteLatency,
+		"xms_read_latency":             data.Content.ReadLatency,
+		"xms_user_accounts_count":      data.Content.NumAccounts,
+	}
+
+	acc.AddFields("xio", fields, tags)
+}
+
+func (xio *XtremIO) call(endpoint string) (string, error) {
+	req, err := http.NewRequest("GET", xio.URL+"/api/json/v3/types/"+endpoint, nil)
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+	req.AddCookie(xio.cookie)
+	resp, err := xio.client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	return string(data), nil
+}
+
+func (xio *XtremIO) authenticate() error {
+	req, err := http.NewRequest("GET", xio.URL+"/api/json/v3/commands/login?password="+xio.Password+"&user="+xio.Username, nil)
```

Comment:
Is this the only way to pass the username/password? Do these devices always use HTTPS?

Reply:
These are generally going to be handled over HTTPS and likely using private IPs, but I guess that decision is ultimately up to whoever is managing the storage array. So it's possible they could be passed over HTTP.

I can also do basic auth, which probably makes more sense than query string params. Let me know if you want me to switch that over to basic auth and I can push the code. I already have it written.

Reply:
@srebhan @reimda thoughts?

Reply:
If basic auth (or even more advanced schemes) is supported, please use that.